File: c++/SCC.cpp
// 強連結成分のグループ番号を返す
// 動作確認未実施

#include <bits/stdc++.h>
using namespace std;

struct StronglyConnectedComponents {
    vector<vector<int>> g, rg;
    vector<int> groupList, rorder;
    vector<int> used;
    int groupNo = 0;

    StronglyConnectedComponents(vector<vector<int>> &_g) {
        int N = _g.size();

        g = _g;
        rg.resize(N);
        groupList.resize(N, -1);
        used.resize(N, false);

        for (int from = 0; from < N; from++) {
            for (int to : g[from]) {
                rg[to].push_back(from);
            }
        }

        for (int i = 0; i < N; i++) {
            if (used[i]) continue;
            dfs(i);
        }

        reverse(rorder.begin(), rorder.end());
        for (int r : rorder) {
            if (groupList[r] != -1) continue;
            rdfs(r);
            groupNo++;
        }
    }

    void dfs(int u) {
        used[u] = true;
        for (int v : g[u]) {
            if (used[v]) continue;
            dfs(v);
        }
        rorder.push_back(u);
    }

    void rdfs(int u) {
        for (int v : rg[u]) {
            if (groupList[v] != -1) continue;
            rdfs(v);
        }
    }

    int group(int i) { return groupList[i]; }
};

File: python/SCC.py
# 強連結成分分解


import sys

sys.setrecursionlimit(10**8)


def scc(N, G):
    used = [False] * N
    order = []

    def dfs(u):
        for v in G[u]:
            if used[v]:
                continue
            used[v] = True
            dfs(v)

        order.append(u)

    for u in range(N):
        if used[u]:
            continue
        used[u] = True
        dfs(u)

    RG = [[] for _ in range(len(G))]

    for i, g in enumerate(G):
        for u in g:
            RG[u].append(i)

    groups = []
    used = [False] * N

    def rdfs(u):
        groups[-1].append(u)

        for v in RG[u]:
            if used[v]:
                continue
            used[v] = True
            rdfs(v)
        return

    while order:
        u = order.pop()

        if used[u]:
            continue
        groups.append(list())

        used[u] = True
        rdfs(u)

    return groups


if __name__ == "__main__":
    # Atcoder Library Practice Contest G-SCC
    input = sys.stdin.readline

    N, M = map(int, input().strip().split())

    G = [[] for _ in range(N)]
    for _ in range(M):
        a, b = map(int, input().strip().split())

        G[a].append(b)

    groups = scc(N, G)

    print(len(groups))
    for group in groups:
        print(len(group), " ".join([str(i) for i in group]))
